---
title: "Introduction to single-cell RNA-seq analysis"
subtitle: "Cluster marker genes - Cell type labelling"
output:
  html_document:
    toc: yes
    toc_float: true
    number_sections: true
    code_folding: show 
    css: ../css/boxes.css
---


```{r setup, echo=FALSE, message=FALSE, warning=FALSE, purl=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, cache=TRUE)
set.seed(123)
```


# Load data

```{r load_packages}
library(scater)
library(scran)
library(tidyverse)
library(patchwork)
```

Re-run `scoreMarkers` on the clustered data.

```{r load_data}
sce <- readRDS("R_objects/Caron_clustered.500.rds")
markers <- scoreMarkers(sce, 
                        groups = sce$label, 
                        block = sce$SampleName)
```

# Cell Type Labelling

One of the main tasks we often want to perform is annotating our cells as known
types of cells ocurring in our sampled tissue. This requires prior knowledge of
cell transcriptomic states and therefore becomes easier if there are
well-curated resources available. However, for less well-studied tissues these
may not be available and so cell type annotation may rely on "manual annotation"
using a small set of genes with known cell-specific expression (e.g. from
microscopy data, qPCR on cell-sorted samples, etc.).

In this section we will do a very simple manual labelling of our clusters, based
on known genes expressed in different blood cell types. However, there are more
systematic methods for cell type annotation, in particular when prior
information is available for those cells:

- The [`SingleR`](https://bioconductor.org/packages/3.16/bioc/vignettes/SingleR/inst/doc/SingleR.html) package uses previously labelled bulk or single-cell datasets to annotate a new dataset.
- The [`AUCcell`](https://bioconductor.org/packages/3.16/bioc/vignettes/AUCell/inst/doc/AUCell.html) package classifies cells into types based on user-provided lists of "signature" genes for each type. These lists can be generated from literature, or also be based on prior RNA-seq datasets. 
- Another strategy is to perform a standard gene set enrichment analysis on the top marker genes for each cluster. 

Any of these approaches is suitable, and they can often be done in conjunction. 
To learn more about these approaches, read the chapter on [cell type annotation in the OSCA book](http://bioconductor.org/books/3.16/OSCA.basic/cell-type-annotation.html).


## Manual annotation

A lot is known about immune cell markers, in particular as many surface markers
have been identified as useful for
[immunophenotyping](https://en.m.wikipedia.org/wiki/Cluster_of_differentiation#Immunophenotyping).

To help us in our annotation, we start by retrieving the top-ranked genes from
each cluster into a list. We do this by looping through the list using the
`lapply()` function and in each case picking the genes with rank < 10 for
Cohen's D statistic:

```{r all_top_markers}
top_markers_all <- lapply(markers, function(x){
  x %>% 
    as.data.frame() %>% 
    filter(rank.logFC.cohen < 10) %>% 
    rownames()
})

top_markers_all
```

If we took some time to examine this list (which would benefit from knowing about the underlying biology of the immune system), we can start to see some genes known to differentiate different types of cells: 

- _HBA1_ and _HBA2_ → expressed in red blood cells (erythrocytes)
- _CST3_ → specific to monocytes
- _CD3E_ and _CD3D_ → specific to T cells
- _NKG7_ → specific to natural killer (NK) T cells
- _CD79A_ and _CD24_ → specific to B cells
- _MS4A1_ (CD20) → a clinically-important antigen used as a target to treat B cell-related diseases, such as leukemia. Cells with this antigen are referred to as CD20+ B cells.

Let's visualise these markers' expression in our clusters:

```{r all_marker_violins}
known_genes <- c(
  "HBA1", # erythrocytes
  "CST3", # monocytes
  "CD3E", # T cells
  "NKG7", # NK T cells
  "CD79A",  # B cells
  "MS4A1" # CD20 B cells
  )

# violin plot
plotExpression(sce, x = "label", features = known_genes)

# scaled heatmap of expression
plotGroupedHeatmap(sce, 
                   features = known_genes,
                   group = "label",
                   block = "SampleGroup", 
                   scale = TRUE, center = TRUE, 
                   zlim = c(-3, 3))
```

Based on the expression patterns for these cells, we classify these clusters as follows: 

| Cluster | Labelled Cell Type                | Evidence      |
| :------ | :-------------------------------- | :------------ |
| 1       | B cells                           | CD79A         |
| 2       | B cells                           | CD79A         |
| 3       | B cells                           | CD79A         |
| 4       | B cells                           | CD79A         |
| 5       | CD20+ B cells                     | CD79A + MS4A1 |
| 6       | T cells                           | CD3D          |
| 7       | NK T cells                        | CD3D + NKG7   |
| 8       | Erythrocytes                      | HBA1          |
| 9       | Erythrocytes                      | HBA1          |
| 10      | Erythrocytes                      | HBA1          |
| 11      | Monocytes                         | CST3          |
| 12      | B cells                           | CD79A         |


Now that we have a more meaningful annotation for our clusters, let's add this to our `SingleCellExperiment` object.
We will also add the original cluster ID in parenthesis to remind ourselves that this annotation was done based on the clusters. 

The cell labels are stored in the `SingleCellExperiment` object as a _factor_ (a type of object in R to store categorical data), and so we can change the labels using the `levels()` function, like so:

```{r relabel}
levels(colLabels(sce)) <- c("B (c1)", "B (c2)", 
                            "B (c3)", "B (c4)",
                            "CD20+ B (c5)", 
                            "T (c6)", "NK T (c7)", 
                            "Erythrocytes (c8)", "Erythrocytes (c9)", 
                            "Erythrocytes c(10)",
                            "Monocytes (c11)", "B (c12)")
```

Now, when we label our UMAP, we can see the new labels, which are more intutitive to interpret:

```{r umap_relabel}
plotReducedDim(sce, dimred = "UMAP_corrected", 
               colour_by = "label", text_by = "label")
```


## (Bonus) Further exploration

### Erythrocytes

Looking at the expression of _HBA1_ and _HBA2_: 

```{r, purl=FALSE}
plotExpression(sce, x = "label", features = c("HBA1", "HBA2")) +
  scale_x_discrete(guide = guide_axis(angle = 45))
```

You may notice that there is a lot of background expression across several cell clusters. 
_HBA1_ and _HBA2_ are common components of the **"soup" or ambient RNA** in scRNA-seq experiments involving blood cells. 
Hemoglobin chains, such as _HBA1_ and _HBA2_, are very highly expressed in red blood cells and burst red blood cells will contribute their RNA to the medium from which the cells are loaded into the 10X Chromium machine. 
Some of this medium, or "soup", containing ambient RNA is captured with each cell. 
There are methods available to correct for soup in your analysis such as [SoupX](https://academic.oup.com/gigascience/article/9/12/giaa151/6049831).


### B Cells

Our simple manual annotation wasn't very conclusive about cluster 12. 
While it expresses the _CD79A_ B-cell marker, it also appears to be quite separate from other B-cell clusters on our UMAP.  

```{r, purl=FALSE}
plotReducedDim(sce, dimred = "UMAP_corrected", 
               colour_by = "CD79A", text_by = "label")
```

This is also the case for cluster 5, however we could see that it likely represents CD20+ B-cells, which explains its separate clustering.
However, we didn't have any specific gene in our small list that distinguishes cluster 12 from the other B-cell clusters. 

We could look at our list of top markers and identify any that are unique to this cluster: 

```{r, purl=FALSE}
setdiff(top_markers_all[[12]], unlist(top_markers_all[1:11]))
```

And use this list to further investigate the specific properties of this cluster. 
For example, we can see that _CD38_ (another "cluster of differentiation" gene) is expressed highly in this cluster: 

```{r, purl=FALSE}
plotExpression(sce, x = "label", 
               features = "CD38") +
  scale_x_discrete(guide = guide_axis(angle = 45))
```

One thing to note is that this cluster only seems to contain cells from PBMMC samples: 

```{r, purl=FALSE}
table(sce$SampleGroup, sce$label)
```

Therefore, it suggests there are absent from our cancer blood samples. 

The difficulty in annotating cluster 12 exemplifies the **limitation of manual cell annotation**, which can become quite laborious and limited in its ability to classify cells. 

Also, we should keep in mind that perhaps our clustering wasn't ideal, there may have been technical issues during normalisation and dataset integration, which we should investigate (e.g. by producing UMAP and clustering from different normalisation methods). 


# Session information

<details><summary>`sessionInfo()`</summary>

```{r, purl=FALSE}
devtools::session_info()
```

</details>
